crates/utils/src/utils/markdown/mod.rs

```diff
@@ -0,0 +1,245 @@
+use crate::settings::SETTINGS;
+use markdown_it::{plugins::cmark::inline::image::Image, MarkdownIt};
+use once_cell::sync::Lazy;
+use url::Url;
+use urlencoding::encode;
+
+mod link_rule;
+mod spoiler_rule;
+
+static MARKDOWN_PARSER: Lazy<MarkdownIt> = Lazy::new(|| {
+  let mut parser = MarkdownIt::new();
+  markdown_it::plugins::cmark::add(&mut parser);
+  markdown_it::plugins::extra::add(&mut parser);
+  spoiler_rule::add(&mut parser);
+  link_rule::add(&mut parser);
+
+  parser
+});
+
+/// Replace special HTML characters in API parameters to prevent XSS attacks.
+///
+/// Taken from https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.md#output-encoding-for-html-contexts
+///
+/// `>` is left in place because it is interpreted as markdown quote.
+pub fn sanitize_html(text: &str) -> String {
+  text
+    .replace('&', "&amp;")
+    .replace('<', "&lt;")
+    .replace('\"', "&quot;")
+    .replace('\'', "&#x27;")
+}
+
+pub fn markdown_to_html(text: &str) -> String {
+  MARKDOWN_PARSER.parse(text).xrender()
+}
+
+/// Rewrites all links to remote domains in markdown, so they go through `/api/v3/image_proxy`.
+pub fn markdown_rewrite_image_links(mut src: String) -> (String, Vec<Url>) {
```

Comment:
I think I commented on this above, but make sure this is only run when you have that proxy_images setting = true.

Reply:
Right that check was missing, added it now.